File: Data_Structure_java/LinkedList/DoublyLinkedList.java
public class DoublyLinkedList<E> {
	private static class Node<E>{
		private E element;
		private Node<E> prev;
		private Node<E> next;
		public Node(E e, Node<E> p, Node<E> n){
			element = e;
			prev = p;
			next = n;
		}
		public E getElement() {return element;}
		public Node<E> getPrev() {return prev;}
		public Node<E> getNext() {return next;}
		public void setPrev(Node<E> p) {prev = p;}
		public void setNext(Node<E> n) {next = n;}
	}

	private Node<E> head;
	public Node<E> tail;
	private int size = 0;
	public DoublyLinkedList(){
		head = new Node<>(null,null,null);
		tail = new Node<>(null, head, null);
		head.setNext(trailer);
	}
	public int size(){return size;}
	public boolean isEmpty(){return size == 0;}
	public E first(){
		if (isEmpty()) return null;
		return head.getNext().getElement();
	}
	public E last(){
		if (isEmpty()) return null;
		return tail.getPrev().getElement();
	}

	public void addFirst(E e){
		addBetween(e, head, head.getNext());
	}
	public void addLast(E e){
		addBetween(e, tail.getPrev(),tail);
	}
	public E removeFirst(){
		if (isEmpty()) return null;
		return remove(head.getNext());
	}
	public E removeLast(){
		if (isEmpty()) return null;
		return remove(tail.getPrev());
	}
	private void addBetween(E e,Node<E> predecessor, Node<E> successor){
		Node<E> newest = new Node<>(e, predecessor, successor);
		predecessor.setNext(newest);
		successor.setPrev(newest);
		size++;
	}
	private E remove(Node<E> node) {
		Node<E> predecessor = node.getPrev();
		Node<E> successor = node.getNext();
		predecessor.setNext(successor);
		successor.setPrev(predecessor);
		size--;
		return node.getElement();
	}
}

File: Data_Structure_python/LinkedList/DoublyLinkedList.py
class Node(object):
	def __init__(self,data,nodeprev,nodenext):
		self.element = data
		self.prev = nodeprev
		self.next = nodenext
	
	def SetPrev(self,nodeprev): 
		self.prev = nodeprev
	def SetNext(self,nodenext): 
		self.next = nodenext
	def getPrev(self):return self.prev
	def getNext(self):return self.next
	def getelement(self) : return self.element

class DoublyLinkedList(Node):
	def __init__(self):
		self.size = 0
		self.header = Node(None,None,None)
		self.trailer = Node(None,None,None)
	def size(self) :return self.size
	def isEmpty(self): return (self.size == 0)
	def addFirst(self,element):
		if self.isEmpty():
			self.header = Node(element,None,None)
			self.trailer = self.header
		else:
			new = Node(element,None,self.header)
			self.header.SetPrev(new)
			self.header = new
		self.size += 1

	def addLast(self,element):
		node = Node(element,None,None) 
		if self.isEmpty():
			self.header = node
		else:
			# node.SetPrev(self.trailer)
			self.trailer.SetNext(node)
		self.trailer = node
		self.size += 1

	def removeFirst(self):
		if self.isEmpty() :return None
		self.remove(self.header)

	def removeLast(self):
		if self.isEmpty() : return None
		self.remove(self.trailer)
	
	def remove(self,node):
		front = node.getPrev()
		last = node.getNext()
		front.SetNext(last)
		last.SetPrev(front)
		self.size -= 1
		return node.getelement()
	
	def printlist(self):
		snake = self.header
		print("Header->",snake.getelement(),end = ' -> ')
		while snake.getNext() != None:
			print(snake.getelement(),end = " -> ")
			snake = snake.getNext()
		print('Trailer')